### ISSUE TITLE
fstab has carriage returns

### ISSUE THREAD

**Issue by remline** [archlinux/archinstall#1399]
After running archinstall, the generated /etc/fstab contains carriage returns (0x0D). This causes the following message in the systemd journal when booting into the newly installed system:

`systemd-fstab-generator[251]: Mount point  is not a valid path, ignoring.`

**Expected behavior:** /etc/fstab should use only line feeds (0x0A) to represent new lines.

This may be related to issue #1378.

**Comment by uhthomas** [comment]
Yeah, I noticed this as well... vim shows `^M` everywhere.

**Comment by kamiccolo** [comment]
short reproducer (with arch-install-scripts cloned as well):

```
from archinstall.lib.general import SysCommand

print(SysCommand("arch-install-scripts/genfstab -pU /").decode())
```

Bare in mind that `genfstab -pU /` does NOT provide ^Ms.

**Comment by kamiccolo** [comment]
I'm almost certain by now it has something  to do with `pty.fork()`.

This (closest to the orginal) do reproduce the issue:

```
import os
import pty
import time

pid, child_fd = pty.fork()

if not pid:
    os.execv('../arch-install-scripts/genfstab', ['-pU', '/'])
else:
    time.sleep(1)
    output = os.read(child_fd, 8192)
    print(output)
```


And this does NOT:
```
import os

output = os.execv('../arch-install-scripts/genfstab', ['-pU', '/'])
print(output)
```

**Comment by codefiles** [comment]
> mhm, `termios.ONLCR`?

@kamiccolo

```python
import os
import pty
import time

pid, fd = pty.fork()
if pid == 0:
    os.execv('/usr/bin/printf', ['printf', 'foo\nbar\n'])
else:
    time.sleep(1)
    output = os.read(fd, 8192)
    print(output)
```

**Output:** `b'foo\r\nbar\r\n'`

Utilizing the `termios` module example [1] to disable the terminal output setting 'translate newline to carriage return-newline' by negation of `onlcr` [2] :

```python
import os
import pty
import termios
import time

pid, fd = pty.fork()
if pid == 0:
    os.execv('/usr/bin/printf', ['printf', 'foo\nbar\n'])
else:
    old = termios.tcgetattr(fd)
    new = termios.tcgetattr(fd)
    new[1] &= ~termios.ONLCR
    try:
        termios.tcsetattr(fd, termios.TCSANOW, new)
        time.sleep(1)
        output = os.read(fd, 8192)
        print(output)
    finally:
        termios.tcsetattr(fd, termios.TCSANOW, old)
```

**Output:** `b'foo\nbar\n'`

[1] https://docs.python.org/3/library/termios.html#example
[2] [https://man.archlinux.org/man/stty.1#Output_settings:](https://man.archlinux.org/man/stty.1#Output_settings:)

**Comment by kamiccolo** [comment]
> 

Woa, way nicer and clearer explanation. Been bashing my head  with those different `termios` flags with not much of a further progress. I wonder, why did this happen in the first place? Some change in underlying vterms? Also, not sure if we should fix this in `SysCommand()`, because it may break some other stuff unrelated to `fstab`. On the other hand, replacing `\r\n` does not look nice as well.

**Comment by Fredolx** [comment]
Is only fstab affected by this?

**Comment by codefiles** [comment]
> But I think setting the tty attribute is the cleanest approach?

Checking the output of the examples below will show that `onlcr` is enabled in both cases but carriage returns are only present in the last example, which is using the `pty` module. Disabling `onlcr` will eliminate the carriage returns when using the `pty` module but it is a workaround.

``` python
import os

output = os.execv('/usr/bin/stty', ['stty', '-a'])
print(output)
```

``` python
import os
import pty
import time

pid, fd = pty.fork()
if pid == 0:
    os.execv('/usr/bin/stty', ['stty', '-a'])
else:
    time.sleep(1)
    output = os.read(fd, 8192)
    print(output)
```

I do not understand why the `subprocess` or `asyncio` module is not being used to start subprocesses. The `subprocess` module was used until commit 1f89be3 but the commit message of "Swapping the run() command for something a little more potent" is obscure.

**Comment by Torxed** [comment]
> "Swapping the run() command for something a little more potent" is obscure.

`subprocess` was not working reliably with `machinectl`, `ssh` or any other process that was forking their `stdin` handlers as well as detecting changes in running state of the executable.
There were probably other reasons too, and those could most likely be worked around with libraries and such, but at the time I felt more intimidated by building around the issues rather than creating a solution myself - without libraries.

But suggestions are always welcome :)

**Comment by codefiles** [comment]
> `subprocess` was not working reliably with `machinectl`, `ssh` or any other process that was forking their `stdin` handlers as well as detecting changes in running state of the executable.

There is not any use of `machinectl` or `ssh` in the code, in what context are they being used? When would "detecting changes in running state of the executable" be required? Could you provide an example that demonstrates the problem?

> But suggestions are always welcome :)

I have none yet because I still lack an understanding of what necessitates the design choices for `archinstall.SysCommand()`.

**Comment by codefiles** [comment]
I understand now that the desire is for `archinstall.SysCommand()` to work with any command and therefore it is designed to use a pseudoterminal. A possible alternative to interacting with `ssh` in this manner would be `asyncssh`.

The only fix other than disabling `onclr` is to use `bytes.replace(b'\r\n', b'\n')`.

```python
import os
import sys
import pty

pid, fd = pty.fork()

if pid == 0:
    os.execv('/usr/bin/printf', ['printf', 'foo\nbar\nbaz\n'])
    sys.exit(1)
else:
    output = b''
    while true:
        try:
            data = os.read(fd, 1024)
            output += data.replace(b'\r\n', b'\n')
        except:
            break
    print(output)
    print(output.decode(), end='')
```

**Output:**

```
b'foo\nbar\nbaz\n'
foo
bar
baz
```

**Comment by kamiccolo** [comment]
I wonder, wasn't this particular issue already solved. There were a few simplifications of `SysCommand` and stuff already.

**Comment by codefiles** [comment]
This issue has not been solved.

The issue was inadvertently fixed for Btrfs in commit 00b0ae7 but the last line of the fstab file will be missing a newline (`\n`).

https://github.com/archlinux/archinstall/blob/f107104103b74ac065a56bd94a0878f0d263e177/archinstall/lib/installer.py#L338

https://github.com/archlinux/archinstall/blob/f107104103b74ac065a56bd94a0878f0d263e177/archinstall/lib/general.py#L433

https://github.com/archlinux/archinstall/blob/f107104103b74ac065a56bd94a0878f0d263e177/archinstall/lib/general.py#L439-L440

The logic added to the `SysCommand.decode()` function in commit 5c903df to use `str.strip()` does not work well considering the output of many commands will be more than one line but that is another issue.

In this for loop if the file system is Btrfs then the fstab file, which will contain the carriage return-newlines (`\r\n`) at this point, is read in with `file.readlines()`. This function removes the carriage returns but leaves the newlines and splits up the lines of the file into a list. Then after potentially manipulating the contents of the list, it is written back to the fstab file with `file.writelines()`.

https://github.com/archlinux/archinstall/blob/f107104103b74ac065a56bd94a0878f0d263e177/archinstall/lib/installer.py#L358-L387

For the intended purpose the read and write of the file only need to occur once if even at all, not iteratively by way of a double nested for loop but that is yet another issue.
